### PR TITLE
SWAT-50: Make it possible to rate-limit sprinkler tasks

### PR DESCRIPTION
**Pull request by apolisskaya**
Rather than having a single task type of "SprinkerBase._async_subtask" in the task registry, let's wrap _async_subtask in a task defined right before creating the main Sprinkler chord. That way, we can set any options we want (retries, rate limit, task name, etc) on subtasks of individual sprinklers to be defined prior to generating instances of those tasks (or proxies).

This change would allow us to modify the subtasks run by the sprinkler for each specific Sprinkler that we use now or may create in the future. It would also make monitoring celery a bit easier, since sprinkler subtasks will now have distinct, descriptive names by default.

Once this is done we can add rate limiting (and any other custom task property we want) onto Sprinkler subtasks like so: https://github.com/groveco/grove/pull/7769/files

Some background info:
- If a Celery task is called directly or with `apply()` rather than calling `<task>.apply_async()` or adding the task to a chord, on which we then call `apply_async` (like in the Sprinklers), the method is called directly rather than being serialized and placed onto the queue. So, it's okay to have a task take another task as its "method". We won't have a task that creates a task.
- Unless we want to configure the subtasks, current sprinklers will continue to function normally when this is merged in. The default proposed behavior, if no options are supplied, mirrors current behavior.
- `task()` takes in some provided method and registers it as a Celery Task. by default, it's registered as its own name, but a custom name can be provided. Please note that creating a differently named task at time of registration is NOT the same thing as changing its `shadow_name` on the task signature, though the effect would be the same from a logging perspective.
- If we try to simply add a rate limit without setting custom task names at registration time, the rate limit will be applied to ALL sprinkler subtasks, which is not desired.

### CODE REVIEW

File: sprinklers/base.py
Comment (Anton-Shutik) [inline sprinklers/base.py]:
```suggestion
    def get_async_subtask_options(self):
```

But I don't feel strongly about it

**Review by Anton-Shutik — COMMENTED** [review]
This looks great! Can we make sure  this works ? Can we add a test for that ? python test can just check that the args are passed to async_task, and we can manually test that rate limit really works on staging

**Comment by apolisskaya** [comment]
@Anton-Shutik I can't get the tests running locally for django-sprinklers, I can test this on staging but will need your help to either write or run the tests for this repo.

I'm actually living in your timezone for the next week and a half, so let's sync up about this!

**Comment by Anton-Shutik** [comment]
@apolisskaya I think testing on staging is fine. You can just make a separate brunch and add TestSprinkler with rate limit. Than run it on staging and watch logs. It will show you how often the sub task executed.

**Comment by apolisskaya** [comment]
@Anton-Shutik Working on that now, see [here](https://github.com/groveco/grove/pull/7772) for fake sprinkler code

File: sprinklers/base.py
Comment (al-the-x) [inline sprinklers/base.py:83]: This was totally @stephen-bunn's crazy idea, FWIW. Blame him.

Comment (apolisskaya) [reply sprinklers/base.py:83]: Does this actually work? Celery needs to be able to import the method for the task

**Comment by daviddahl** [comment]
I would love to see a unit test that just checks the logic we have added here. Nothing fancy, just make sure `legacy` subtasks are handled correctly.

File: sprinklers/base.py
Comment (stephen-bunn) [reply sprinklers/base.py:83]: It should. As long as `current_app.task` has been called during import on the `_async_subtask` callable (whether it's defined in the sprinkler or in the sprinkler base), celery should have already registered the task and stored the callable.

**Comment by al-the-x** [comment]
@Anton-Shutik or @chrisclark would appreciate your 👍 (and approved review) on this so we can unblock groveco/grove#7769

**Review by Anton-Shutik — APPROVED** [review]
I've made small [branch](https://github.com/groveco/grove/compare/play-with-sprinklers?expand=1) and played it on staging2. Run small script from the shell:
`heroku run python manage.py shell -a protopantry2`
```python
from core.cache import get_redis
from membership.tasks import DummySprinkler, SimpleSprinkler

# remove previously added tasks
r = get_redis()
r.flushdb()

SimpleSprinkler().start()
# waited until it finished and the run:
DummySprinkler().start()

```
and in parallel watched logs from worker, who handles default queue:
`heroku logs -t -a protopantry2 -d worker`

I was able to see what I expected:
1. 1st sprinkler finished almost immediately 
2. 2nd processed one task per 6 seconds

LGTM, but also would like another set of eyes on this PR

**Comment by al-the-x** [comment]
I observed the same in my staging tests with groveco/grove#7772. Merging this down so that @apolisskaya can update her PR.

**Comment by apolisskaya** [comment]
@al-the-x thank you so much for picking this up, updating my PR now to get our Membershipbot and Chargebot sprinklers slowed down!

**Pull request merged**
